File: languages/php/src/BitwardenLib.php
<?php

namespace Bitwarden\Sdk;

use Bitwarden\Sdk\Schemas\ClientSettings;
use Bitwarden\Sdk\Schemas\Command;
use Exception;
use FFI;
use JsonException;
use RuntimeException;
use stdClass;

class BitwardenLib
{
    public FFI $ffi;
    public FFI\CData $handle;

    /**
     * @throws Exception
     */
    public function __construct()
    {
        $lib_file = null;

        if (PHP_OS === 'WINNT') {
            $lib_file = __DIR__ . '/lib/windows-x64/bitwarden_c.dll';
            if (!file_exists($lib_file)) {
                $lib_file = __DIR__ . '/../../../target/debug/bitwarden_c.dll';
            }
        } elseif (PHP_OS === 'Linux') {
            $lib_file = __DIR__ . '/lib/linux-x64/libbitwarden_c.so';
            if (!file_exists($lib_file)) {
                $lib_file = __DIR__ . '/../../../target/debug/libbitwarden_c.so';
            }
        } elseif (PHP_OS === 'Darwin') {
            $architecture = trim(exec('uname -m'));
            if ($architecture === 'x86_64' || $architecture === 'amd64') {
                $lib_file = __DIR__ . '/lib/macos-x64/libbitwarden_c.dylib';
            } elseif ($architecture === 'arm64') {
                $lib_file = __DIR__ . '/lib/macos-arm64/libbitwarden_c.dylib';
            }
            if (!file_exists($lib_file)) {
                $lib_file = __DIR__ . '/../../../target/debug/libbitwarden_c.dylib';
            }
        }

        if ($lib_file == null || !is_file($lib_file)) {
            throw new Exception("Lib file not found");
        }

        $this->ffi = FFI::cdef('
            void* init(const char* param);
            char* run_command(void* c_str_ptr, void* client_ptr);
            void free_mem(void* client_ptr);',
            $lib_file
        );
    }

    /**
     * @throws JsonException
     * @throws Exception
     */
    public function init(ClientSettings $client_settings): FFI\CData
    {
        $encoded_json = $this::json_encode_sdk_format($client_settings->to());
        $this->handle = $this->ffi->init($encoded_json);
        return $this->handle;
    }

    /**
     * @throws JsonException
     * @throws Exception
     */
    public function run_command(Command $command): stdClass
    {
        $encoded_json = $this::json_encode_sdk_format($command->to());
        try {
            $result = $this->ffi->run_command($encoded_json, $this->handle);
            return json_decode(FFI::string($result));
        } catch (FFI\Exception $e) {
            throw new RuntimeException('Error occurred during FFI operation: ' . $e->getMessage());
        }
    }

    public function free_mem(): void
    {
        $this->ffi->free_mem($this->handle);
    }

    /**
     * @throws JsonException
     */
    private static function json_encode_sdk_format($object): string
    {
        $withoutNull = function ($a) use (&$withoutNull) {
            if (is_object($a)) {
                $a = array_filter((array)$a);
                foreach ($a as $k => $v) {
                    $a[$k] = $withoutNull($v);
                }

                return (object)$a;
            }

            return $a;
        };

        $object_no_nulls = $withoutNull($object);

        return json_encode($object_no_nulls, JSON_THROW_ON_ERROR | JSON_UNESCAPED_SLASHES);
    }
}

File: languages/ruby/bitwarden_sdk/lib/bitwarden_lib.rb
# frozen_string_literal: true

require 'ffi'

module BitwardenSDK
  module BitwardenLib
    extend FFI::Library

    def self.mac_with_intel?
      `uname -m`.strip == 'x86_64'
    end

    ffi_lib case RUBY_PLATFORM
            when /darwin/
              local_file = if mac_with_intel?
                             File.expand_path('macos-x64/libbitwarden_c.dylib', __dir__)
                           else
                             File.expand_path('macos-arm64/libbitwarden_c.dylib', __dir__)
                           end
              File.exist?(local_file) ? local_file : File.expand_path('../../../../target/debug/libbitwarden_c.dylib', __dir__)
            when /linux/
              local_file = File.expand_path('linux-x64/libbitwarden_c.so', __dir__)
              File.exist?(local_file) ? local_file : File.expand_path('../../../../target/debug/libbitwarden_c.so', __dir__)
            when /mswin|mingw/
              local_file = File.expand_path('windows-x64/bitwarden_c.dll', __dir__)
              File.exist?(local_file) ? local_file : File.expand_path('../../../../target/debug/bitwarden_c.dll', __dir__)
            else
              raise "Unsupported platform: #{RUBY_PLATFORM}"
            end

    attach_function :init, [:string], :pointer
    attach_function :run_command, %i[string pointer], :string
    attach_function :free_mem, [:pointer], :void
  end
end
